src/doodle/doodleCanvas.js

```diff
@@ -31,6 +31,28 @@ const DoodleCanvas = ({
   setLines,
 }) => {
   const [isDrawing, setIsDrawing] = useState(false);
+  const [everActive, setEverActive] = useState(false);
+
+  if (active && !everActive) {
+    setEverActive(true);
+  }
+
+  useEffect(() => {
+    if (lines.length === 0) {
+      return () => {};
+    }
+    const warn = e => {
+      e = e || window.event;
+
+      e.preventDefault();
+      e.returnValue = '';
+      return '';
+    };
+    window.addEventListener('beforeunload', warn);
+    return () => {
+      window.removeEventListener('beforeunload', warn);
+    };
```

Comment:
Can you explain what is going on here? Confused on the adding and removal of event listener and the return value of warn.

Reply:
So warn is the beforeunload event listener that's supposed to give the unsaved changes message. Because browsers respond in different ways, either e.preventDefaunt(), e.returnValue = <not undefined>, or return <not undefined> will cause this message to appear when leaving the page. For cross-browser compatibility, all three are used.

As for adding/removing the listener, that's to get it to only fire once, and only if they actually have unsaved changes. This is using syntax provided by the `useEffect` hook. See https://reactjs.org/docs/hooks-effect.html#effects-with-cleanup. Effectively, every time `lines` changes, the old listener is removed, and then if lines.length is nonzero, a new listener is added. So when the component is removed, or when lines is cleared, the listener disappears and users will no longer be notified of unsaved changes (because there aren't any). 

I hope that made sense